In [1]:
import pandas as pd

# Data Exploration

In [2]:
review = pd.read_csv('googleplaystore_user_reviews.csv')

In [3]:
review.head()

,App,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity
0,10 Best Foods for You,I like eat delicious food. That's I'm cooking ...,Positive,1.00,0.533333
1,10 Best Foods for You,This help eating healthy exercise regular basis,Positive,0.25,0.288462
2,10 Best Foods for You,NaN,NaN,NaN,NaN
3,10 Best Foods for You,Works great especially going grocery store,Positive,0.40,0.875000
4,10 Best Foods for You,Best idea us,Positive,1.00,0.300000


In [4]:
playstore = pd.read_csv('googleplaystore.csv')

In [5]:
playstore.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up


In [6]:
review.shape

(64295, 5)

In [7]:
playstore.shape

(10841, 13)

In [8]:
review.info()

<class 'pandas.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  str    
 1   Translated_Review       37427 non-null  str    
 2   Sentiment               37432 non-null  str    
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), str(3)
memory usage: 2.5 MB


In [9]:
playstore.info()

<class 'pandas.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  str    
 1   Category        10841 non-null  str    
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  str    
 4   Size            10841 non-null  str    
 5   Installs        10841 non-null  str    
 6   Type            10840 non-null  str    
 7   Price           10841 non-null  str    
 8   Content Rating  10840 non-null  str    
 9   Genres          10841 non-null  str    
 10  Last Updated    10841 non-null  str    
 11  Current Ver     10833 non-null  str    
 12  Android Ver     10838 non-null  str    
dtypes: float64(1), str(12)
memory usage: 1.1 MB


In [10]:
review.columns

Index(['App', 'Translated_Review', 'Sentiment', 'Sentiment_Polarity',
       'Sentiment_Subjectivity'],
      dtype='str')

In [11]:
playstore.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='str')

In [12]:
review.describe()

,Sentiment_Polarity,Sentiment_Subjectivity
count,37432.000000,37432.000000
mean,0.182146,0.492704
std,0.351301,0.259949
min,-1.000000,0.000000
25%,0.000000,0.357143
50%,0.150000,0.514286
75%,0.400000,0.650000
max,1.000000,1.000000


In [13]:
playstore.describe()

,Rating
count,9367.000000
mean,4.193338
std,0.537431
min,1.000000
25%,4.000000
50%,4.300000
75%,4.500000
max,19.000000


# Handle Missing

The maximum rating is 19.0, which is outside the valid rating range (1–5). This indicates a data quality issue and the row should be examined before calculating statistics such as the mean or filling missing values.

In [14]:
playstore[playstore['Rating'] > 5]

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
10472,Life Made WI-Fi Touchscreen Photo Frame,1.9,19.0,3.0M,"1,000+",Free,0,Everyone,NaN,"February 11, 2018",1.0.19,4.0 and up,NaN


In [15]:
playstore.iloc[1]

App                     Coloring book moana
Category                     ART_AND_DESIGN
Rating                                  3.9
Reviews                                 967
Size                                    14M
Installs                           500,000+
Type                                   Free
Price                                     0
Content Rating                     Everyone
Genres            Art & Design;Pretend Play
Last Updated               January 15, 2018
Current Ver                           2.0.0
Android Ver                    4.0.3 and up
Name: 1, dtype: object

**Data Quality Issue**

An invalid record was identified where the columns were shifted, causing incorrect values such as:

- Category = 1.9
- Rating = 19.0

Since Google Play Store ratings must be between 1 and 5 and categories should contain text values, this row was determined to be corrupted and was removed from the dataset before further analysis.

In [16]:
playstore = playstore.drop(10472)

In [17]:
playstore.describe()

,Rating
count,9366.000000
mean,4.191757
std,0.515219
min,1.000000
25%,4.000000
50%,4.300000
75%,4.500000
max,5.000000


In [18]:
playstore.isna().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          8
Android Ver          2
dtype: int64

In [19]:
playstore['Type'].value_counts()

Type
Free    10039
Paid      800
Name: count, dtype: int64

In [20]:
playstore[playstore['Type'].isna() == True]

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
9148,Command & Conquer: Rivals,FAMILY,NaN,0,Varies with device,0,NaN,0,Everyone 10+,Strategy,"June 28, 2018",Varies with device,Varies with device


In [21]:
playstore = playstore.dropna(subset=['Type', 'Current Ver', 'Android Ver'])

In [22]:
playstore.isna().sum()

App                  0
Category             0
Rating            1469
Reviews              0
Size                 0
Installs             0
Type                 0
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          0
Android Ver          0
dtype: int64

In [23]:
playstore.shape

(10829, 13)

In [24]:
playstore['Rating'] = playstore['Rating'].fillna(playstore['Rating'].mean())

In [25]:
playstore.isna().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
dtype: int64

In [26]:
playstore.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up


In [27]:
playstore.dtypes

App                   str
Category              str
Rating            float64
Reviews               str
Size                  str
Installs              str
Type                  str
Price                 str
Content Rating        str
Genres                str
Last Updated          str
Current Ver           str
Android Ver           str
dtype: object

Now let's do the convertions

# Data Cleaning



##### Installs column str -> int

In [28]:
playstore['Installs'].str.endswith('+').sum()

np.int64(10829)

In [29]:
playstore.shape

(10829, 13)

In [30]:
playstore['Installs_clean'] = playstore['Installs'].str.replace("+","")

In [31]:
playstore.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,Installs_clean
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up,"10,000"
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,"500,000"
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up,"5,000,000"
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up,"50,000,000"
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up,"100,000"


In [32]:
playstore['Installs_clean'] = playstore['Installs_clean'].str.replace(",","")

In [33]:
playstore.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,Installs_clean
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up,10000
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,500000
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up,5000000
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up,50000000
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up,100000


In [34]:
playstore['Installs_clean'] = playstore['Installs_clean'].astype(int)

In [35]:
playstore.dtypes

App                   str
Category              str
Rating            float64
Reviews               str
Size                  str
Installs              str
Type                  str
Price                 str
Content Rating        str
Genres                str
Last Updated          str
Current Ver           str
Android Ver           str
Installs_clean      int64
dtype: object

##### Size column str -> int 

In [36]:
playstore['Size'].str.endswith('M').sum()

np.int64(8823)

In [37]:
playstore['Size'].str.endswith('k').sum()

np.int64(312)

In [38]:
print(8823  + 312)

9135


In [39]:
playstore.shape

(10829, 14)

In [40]:
playstore.sample(10)

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,Installs_clean
6466,BM Box,BOOKS_AND_REFERENCE,4.300000,28,Varies with device,"1,000+",Free,0,Everyone,Books & Reference,"October 20, 2016",Varies with device,Varies with device,1000
3409,icon wallpaper dressup💞CocoPPa,PERSONALIZATION,4.300000,595120,Varies with device,"10,000,000+",Free,0,Everyone,Personalization,"July 10, 2018",Varies with device,Varies with device,10000000
6843,Bu Hangi Film ?,FAMILY,4.191838,3,25M,100+,Free,0,Everyone,Puzzle,"August 31, 2017",3.2.2dk,4.0.3 and up,100
7509,App for Craigslist Pro - Buy & Sell Postings,LIFESTYLE,4.300000,707,5.0M,"100,000+",Free,0,Everyone,Lifestyle,"May 9, 2018",App copyright,4.1 and up,100000
7775,Servidor Privado CR y CoC - Royale Servers PRO,TOOLS,4.191838,55,4.3M,100+,Paid,$1.99,Everyone,Tools,"December 24, 2017",1.0.3,2.3 and up,100
1234,Instacart: Grocery Delivery,FOOD_AND_DRINK,4.400000,17071,Varies with device,"1,000,000+",Free,0,Everyone,Food & Drink,"August 2, 2018",Varies with device,5.0 and up,1000000
3798,Updates for Samsung - Android Update Versions,NEWS_AND_MAGAZINES,4.000000,80368,Varies with device,"10,000,000+",Free,0,Everyone,News & Magazines,"May 17, 2018",Varies with device,Varies with device,10000000
2082,Montessori Preschool,FAMILY,3.900000,237,26M,"100,000+",Free,0,Everyone,Education;Education,"July 12, 2018",1.7.2,4.1 and up,100000
4536,Offroad Pickup Truck R,FAMILY,4.400000,6367,96M,"1,000,000+",Free,0,Everyone,Simulation,"July 14, 2017",1.4,2.3 and up,1000000
3136,TripAdvisor Hotels Flights Restaurants Attract...,TRAVEL_AND_LOCAL,4.400000,1162838,Varies with device,"100,000,000+",Free,0,Everyone,Travel & Local,"August 4, 2018",Varies with device,Varies with device,100000000


In [41]:
playstore[playstore['Size'] == 'Varies with device']

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,Installs_clean
37,Floor Plan Creator,ART_AND_DESIGN,4.1,36639,Varies with device,"5,000,000+",Free,0,Everyone,Art & Design,"July 14, 2018",Varies with device,2.3.3 and up,5000000
42,Textgram - write on photos,ART_AND_DESIGN,4.4,295221,Varies with device,"10,000,000+",Free,0,Everyone,Art & Design,"July 30, 2018",Varies with device,Varies with device,10000000
52,Used Cars and Trucks for Sale,AUTO_AND_VEHICLES,4.6,17057,Varies with device,"1,000,000+",Free,0,Everyone,Auto & Vehicles,"July 30, 2018",Varies with device,Varies with device,1000000
67,Ulysse Speedometer,AUTO_AND_VEHICLES,4.3,40211,Varies with device,"5,000,000+",Free,0,Everyone,Auto & Vehicles,"July 30, 2018",Varies with device,Varies with device,5000000
68,REPUVE,AUTO_AND_VEHICLES,3.9,356,Varies with device,"100,000+",Free,0,Everyone,Auto & Vehicles,"May 25, 2018",Varies with device,Varies with device,100000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10713,My Earthquake Alerts - US & Worldwide Earthquakes,WEATHER,4.4,3471,Varies with device,"100,000+",Free,0,Everyone,Weather,"July 24, 2018",Varies with device,Varies with device,100000
10725,Posta App,MAPS_AND_NAVIGATION,3.6,8,Varies with device,"1,000+",Free,0,Everyone,Maps & Navigation,"September 27, 2017",Varies with device,4.4 and up,1000
10765,Chat For Strangers - Video Chat,SOCIAL,3.4,622,Varies with device,"100,000+",Free,0,Mature 17+,Social,"May 23, 2018",Varies with device,Varies with device,100000
10826,Frim: get new friends on local chat rooms,SOCIAL,4.0,88486,Varies with device,"5,000,000+",Free,0,Mature 17+,Social,"March 23, 2018",Varies with device,Varies with device,5000000


In [42]:
print(8823  + 312 + 1694)

10829


In [43]:
playstore.shape

(10829, 14)

let's keep the size column as it is for now

##### Price column str -> int

In [44]:
playstore['Price'].str.startswith('$').sum()

np.int64(797)

In [45]:
playstore['Price'].str.startswith('0').sum()

np.int64(10032)

In [46]:
print(797 + 10032)

10829


In [47]:
playstore['Price_clean'] = playstore['Price'].str.replace('$','')

In [48]:
playstore['Price_clean'].str.startswith('$').sum()

np.int64(0)

In [49]:
playstore['Price_clean'] = playstore['Price_clean'].astype(float)

In [50]:
playstore.dtypes

App                   str
Category              str
Rating            float64
Reviews               str
Size                  str
Installs              str
Type                  str
Price                 str
Content Rating        str
Genres                str
Last Updated          str
Current Ver           str
Android Ver           str
Installs_clean      int64
Price_clean       float64
dtype: object

In [51]:
playstore.sample(10)

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,Installs_clean,Price_clean
3327,Smart Keyboard Trial,TOOLS,4.100000,65597,2.4M,"10,000,000+",Free,0,Everyone,Tools,"April 28, 2017",4.20.1,2.3 and up,10000000,0.00
6020,BD All Sim Offer,BOOKS_AND_REFERENCE,4.500000,197,2.8M,"10,000+",Free,0,Everyone,Books & Reference,"January 3, 2017",1.2,4.1 and up,10000,0.00
6728,BS Meter (Ad Supported),FAMILY,3.900000,14,592k,"1,000+",Free,0,Teen,Entertainment,"August 14, 2013",2.0,2.2 and up,1000,0.00
6962,Kick Axe Bx,HEALTH_AND_FITNESS,4.191838,0,9.3M,1+,Free,0,Everyone,Health & Fitness,"June 7, 2017",1.0.1,4.0 and up,1,0.00
6440,MEGA MAN MOBILE,GAME,4.000000,267,40M,"10,000+",Paid,$1.99,Everyone,Action,"February 17, 2017",1.02.01,4.0 and up,10000,1.99
3205,"KAYAK Flights, Hotels & Cars",TRAVEL_AND_LOCAL,4.500000,216388,19M,"10,000,000+",Free,0,Everyone,Travel & Local,"July 31, 2018",Varies with device,Varies with device,10000000,0.00
1169,Capital One® Mobile,FINANCE,4.600000,510401,79M,"10,000,000+",Free,0,Everyone,Finance,"August 1, 2018",5.38.1,5.0 and up,10000000,0.00
10057,GLASS GO Launcher EX Theme,PERSONALIZATION,4.200000,18280,6.0M,"1,000,000+",Free,0,Everyone,Personalization,"January 8, 2015",1.28,1.6 and up,1000000,0.00
232,Fast Scanner : Free PDF Scan,BUSINESS,4.500000,103755,14M,"10,000,000+",Free,0,Everyone,Business,"July 11, 2018",3.9.2,4.1 and up,10000000,0.00
7655,Central Bank and Trust Co.,FINANCE,4.700000,65,12M,"5,000+",Free,0,Everyone,Finance,"June 12, 2018",5.9.1.0,5.0 and up,5000,0.00


##### Reviews column str -> int

In [52]:
playstore['Reviews_clean'] = playstore['Reviews'].astype(int)

In [53]:
playstore.dtypes

App                   str
Category              str
Rating            float64
Reviews               str
Size                  str
Installs              str
Type                  str
Price                 str
Content Rating        str
Genres                str
Last Updated          str
Current Ver           str
Android Ver           str
Installs_clean      int64
Price_clean       float64
Reviews_clean       int64
dtype: object

In [54]:
playstore.describe()

,Rating,Installs_clean,Price_clean,Reviews_clean
count,10829.000000,1.082900e+04,10829.000000,1.082900e+04
mean,4.191838,1.547990e+07,1.028091,4.446018e+05
std,0.479038,8.507114e+07,15.957778,2.929213e+06
min,1.000000,0.000000e+00,0.000000,0.000000e+00
25%,4.100000,5.000000e+03,0.000000,3.800000e+01
50%,4.200000,1.000000e+05,0.000000,2.100000e+03
75%,4.500000,5.000000e+06,0.000000,5.481500e+04
max,5.000000,1.000000e+09,400.000000,7.815831e+07


In [55]:
playstore.to_csv('new_googleplaystore.csv')

# Analysis

In [56]:
playstore["Category"].value_counts()

Category
FAMILY                 1968
GAME                   1144
TOOLS                   841
MEDICAL                 463
BUSINESS                460
PRODUCTIVITY            424
PERSONALIZATION         390
COMMUNICATION           387
SPORTS                  384
LIFESTYLE               382
FINANCE                 366
HEALTH_AND_FITNESS      341
PHOTOGRAPHY             335
SOCIAL                  295
NEWS_AND_MAGAZINES      283
SHOPPING                260
TRAVEL_AND_LOCAL        258
DATING                  234
BOOKS_AND_REFERENCE     230
VIDEO_PLAYERS           175
EDUCATION               156
ENTERTAINMENT           149
MAPS_AND_NAVIGATION     137
FOOD_AND_DRINK          127
HOUSE_AND_HOME           88
AUTO_AND_VEHICLES        85
LIBRARIES_AND_DEMO       84
WEATHER                  82
ART_AND_DESIGN           64
EVENTS                   64
COMICS                   60
PARENTING                60
BEAUTY                   53
Name: count, dtype: int64

By default, value_counts() returns counts in descending order of frequency.

1. Most common category is FAMILY        

2. Top 10 categories are:
    1. Family
    2. Game
    3. Tools
    4. Medical
    5. Business
    6. Productivity
    7. Personalization
    8. Communication
    9. Sports
    10. Lifestyle         

Highest rated categories

In [57]:
playstore.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver', 'Installs_clean', 'Price_clean', 'Reviews_clean'],
      dtype='str')

In [58]:
pd.pivot_table(playstore,index='Category',values='Rating',aggfunc='mean').sort_values(by='Rating',ascending=False).reset_index()

,Category,Rating
0,EDUCATION,4.387768
1,ART_AND_DESIGN,4.368367
2,EVENTS,4.363202
3,BOOKS_AND_REFERENCE,4.311198
4,PERSONALIZATION,4.306573
5,GAME,4.282444
6,PARENTING,4.281973
7,HEALTH_AND_FITNESS,4.266102
8,BEAUTY,4.260570
9,SHOPPING,4.253925


EDUCATION has the highest average rating.

In [59]:
pd.pivot_table(playstore,index='Category',values='Installs_clean',aggfunc='sum').sort_values(by='Installs_clean',ascending=False).reset_index()

,Category,Installs_clean
0,GAME,35086024415
1,COMMUNICATION,32647276251
2,PRODUCTIVITY,14176091369
3,SOCIAL,14069867902
4,TOOLS,11452271905
5,FAMILY,10258203405
6,PHOTOGRAPHY,10088247655
7,NEWS_AND_MAGAZINES,7496317760
8,TRAVEL_AND_LOCAL,6868887146
9,VIDEO_PLAYERS,6222002720


In [60]:
playstore.groupby("Category")["Installs_clean"].sum().sort_values(ascending=False)

Category
GAME                   35086024415
COMMUNICATION          32647276251
PRODUCTIVITY           14176091369
SOCIAL                 14069867902
TOOLS                  11452271905
FAMILY                 10258203405
PHOTOGRAPHY            10088247655
NEWS_AND_MAGAZINES      7496317760
TRAVEL_AND_LOCAL        6868887146
VIDEO_PLAYERS           6222002720
SHOPPING                3247848785
ENTERTAINMENT           2869160000
PERSONALIZATION         2325483782
BOOKS_AND_REFERENCE     1921468576
SPORTS                  1751174498
HEALTH_AND_FITNESS      1583072512
BUSINESS                1001914865
FINANCE                  876648734
EDUCATION                871452000
MAPS_AND_NAVIGATION      724281890
LIFESTYLE                537643539
WEATHER                  426100520
FOOD_AND_DRINK           273898751
DATING                   264310807
HOUSE_AND_HOME           168712461
ART_AND_DESIGN           124333100
LIBRARIES_AND_DEMO        61995910
COMICS                    56086150
MEDICAL    

GAME category has the most installs

Free vs Paid

In [61]:
playstore.groupby("Type").agg({"Rating":"mean","Installs_clean":"sum"})

,Rating,Installs_clean
Type,,
Free,4.187017,167558951306
Paid,4.252521,72905071


- Do free apps have more installs? yes
- Do paid apps have higher ratings? yes

In [62]:
playstore.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,Installs_clean,Price_clean,Reviews_clean
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up,10000,0.0,159
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,500000,0.0,967
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up,5000000,0.0,87510
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up,50000000,0.0,215644
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up,100000,0.0,967


In [63]:
pd.pivot_table(playstore,index='Category',values='Rating',aggfunc=['mean','count']).sort_values(by=('count', 'Rating'), ascending=False)

,mean,count
,Rating,Rating
Category,,
FAMILY,4.192321,1968
GAME,4.282444,1144
TOOLS,4.065777,841
MEDICAL,4.189801,463
BUSINESS,4.145475,460
PRODUCTIVITY,4.208029,424
PERSONALIZATION,4.306573,390
COMMUNICATION,4.163613,387


Another method

In [64]:
pt = pd.pivot_table(
    playstore,
    index='Category',
    values='Rating',
    aggfunc=['mean', 'count']
)

pt.columns = ['mean_rating', 'rating_count']

pt.sort_values('mean_rating', ascending=False)

,mean_rating,rating_count
Category,,
EDUCATION,4.387768,156
ART_AND_DESIGN,4.368367,64
EVENTS,4.363202,64
BOOKS_AND_REFERENCE,4.311198,230
PERSONALIZATION,4.306573,390
GAME,4.282444,1144
PARENTING,4.281973,60
HEALTH_AND_FITNESS,4.266102,341
BEAUTY,4.260570,53


Category vs Type

In [65]:
pd.pivot_table(playstore,index='Category',columns='Type',values='Rating', aggfunc='mean',fill_value=0)

Type,Free,Paid
Category,,
ART_AND_DESIGN,4.350418,4.733333
AUTO_AND_VEHICLES,4.185590,4.327892
BEAUTY,4.260570,0.000000
BOOKS_AND_REFERENCE,4.324450,4.215598
BUSINESS,4.143818,4.198251
COMICS,4.156395,0.000000
COMMUNICATION,4.169331,4.087377
DATING,4.011921,3.867930
EDUCATION,4.378236,4.750000


# Reviews Dataset

In [66]:
review.head()

,App,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity
0,10 Best Foods for You,I like eat delicious food. That's I'm cooking ...,Positive,1.00,0.533333
1,10 Best Foods for You,This help eating healthy exercise regular basis,Positive,0.25,0.288462
2,10 Best Foods for You,NaN,NaN,NaN,NaN
3,10 Best Foods for You,Works great especially going grocery store,Positive,0.40,0.875000
4,10 Best Foods for You,Best idea us,Positive,1.00,0.300000


In [67]:
review.info()

<class 'pandas.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  str    
 1   Translated_Review       37427 non-null  str    
 2   Sentiment               37432 non-null  str    
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), str(3)
memory usage: 2.5 MB


In [68]:
review["Sentiment"].value_counts()

Sentiment
Positive    23998
Negative     8271
Neutral      5163
Name: count, dtype: int64

- How many Positive reviews? 23998
- How many Negative reviews? 8271
- How many Neutral reviews? 5163

In [69]:
review.groupby('App').agg({'Sentiment_Polarity':'mean'})

,Sentiment_Polarity
App,
10 Best Foods for You,0.470733
104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室,0.392405
11st,0.181294
1800 Contacts - Lens Store,0.318145
1LINE – One Line with One Touch,0.196290
...,...
Hotspot Shield Free VPN Proxy & Wi-Fi Security,0.251765
Hotstar,0.038178
Hotwire Hotel & Car Rental App,0.187029


# Merge

In [70]:
playstore_reviews = pd.merge(playstore,review,on='App',how='left')

In [71]:
playstore_reviews.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,Installs_clean,Price_clean,Reviews_clean,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up,10000,0.0,159,NaN,NaN,NaN,NaN
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,500000,0.0,967,A kid's excessive ads. The types ads allowed a...,Negative,-0.250,1.000000
2,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,500000,0.0,967,It bad >:(,Negative,-0.725,0.833333
3,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,500000,0.0,967,like,Neutral,0.000,0.000000
4,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,500000,0.0,967,NaN,NaN,NaN,NaN


In [72]:
playstore_reviews.shape

(131959, 20)

In [73]:
playstore_reviews.info()

<class 'pandas.DataFrame'>
RangeIndex: 131959 entries, 0 to 131958
Data columns (total 20 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   App                     131959 non-null  str    
 1   Category                131959 non-null  str    
 2   Rating                  131959 non-null  float64
 3   Reviews                 131959 non-null  str    
 4   Size                    131959 non-null  str    
 5   Installs                131959 non-null  str    
 6   Type                    131959 non-null  str    
 7   Price                   131959 non-null  str    
 8   Content Rating          131959 non-null  str    
 9   Genres                  131959 non-null  str    
 10  Last Updated            131959 non-null  str    
 11  Current Ver             131959 non-null  str    
 12  Android Ver             131959 non-null  str    
 13  Installs_clean          131959 non-null  int64  
 14  Price_clean             131959 

# Additional Questions

1. Do higher-rated apps have more positive reviews?
2. Which category has the most positive sentiment?

For each app, what is the average rating associated with each sentiment?


In [81]:
pd.pivot_table(playstore_reviews,index='App',values='Rating',columns='Sentiment',aggfunc='mean')

Sentiment,Negative,Neutral,Positive
App,,,
10 Best Foods for You,4.0,4.0,4.0
11st,3.8,3.8,3.8
1800 Contacts - Lens Store,4.7,4.7,4.7
1LINE – One Line with One Touch,4.6,4.6,4.6
2018Emoji Keyboard 😂 Emoticons Lite -sticker&gif,4.2,4.2,4.2
...,...,...,...
Hotels.com: Book Hotel Rooms & Find Vacation Deals,4.5,4.5,4.5
Hotspot Shield Free VPN Proxy & Wi-Fi Security,4.2,4.2,4.2
Hotstar,4.3,4.3,4.3


### Output

| App   | Negative | Positive |
|--------|----------|----------|
| App A | 4.2 | 4.2 |
| App B | 3.8 | 3.8 |
| App C | 4.6 | 4.6 |

### Notice something interesting

For a given app, the rating is usually the same for every review of that app.

If **App A** has a rating of **4.2**, then:

- Positive reviews → **4.2**
- Negative reviews → **4.2**
- Neutral reviews → **4.2**

This happens because the **rating belongs to the app**, not to the individual review.

Therefore, this pivot table often doesn't provide much additional insight, since the average rating remains the same across different sentiment categories for the same app.

**Summary**
- Rating is given by the app store (overall score)
- Sentiment is derived from user review text

So that doesn't give you the answer to the "Do higher-rated apps have more positive reviews?" question

In [108]:
pd.crosstab(playstore_reviews['App'],playstore_reviews['Sentiment']).sort_values(by='Positive',ascending=False)

Sentiment,Negative,Neutral,Positive
App,,,
Duolingo: Learn Languages Free,238,42,1400
Helix Jump,348,36,1254
Bowmasters,714,144,1014
"CBS Sports App - Scores, News, Stats & Watch Live",336,216,984
Calorie Counter - MyFitnessPal,260,170,845
...,...,...,...
Draw In,1,0,0
Discover Mobile,1,0,0
HD Camera,0,1,0


Since Now the 'App' is the index i need to reset_index()

In [ ]:
pd.crosstab(playstore_reviews['App'],playstore_reviews['Sentiment']).reset_index()

Sentiment,App,Negative,Neutral,Positive
0,10 Best Foods for You,20,44,324
1,11st,7,10,23
2,1800 Contacts - Lens Store,12,20,128
3,1LINE – One Line with One Touch,8,3,27
4,2018Emoji Keyboard 😂 Emoticons Lite -sticker&gif,1,6,25
...,...,...,...,...
811,Hotels.com: Book Hotel Rooms & Find Vacation D...,42,16,78
812,Hotspot Shield Free VPN Proxy & Wi-Fi Security,3,14,17
813,Hotstar,14,4,14
814,Hotwire Hotel & Car Rental App,6,11,16


In [110]:
sentiment_counts = pd.crosstab(playstore_reviews['App'],playstore_reviews['Sentiment']).reset_index()

In [106]:
pd.pivot_table(playstore_reviews,index='App',values='Rating',aggfunc='mean').sort_values(by='Rating',ascending=False)

,Rating
App,
HON. B.J. ACS COLLEGE ALE,5.0
iReadMe,5.0
211:CK,5.0
meStudying: AP English Lit,5.0
Hey AJ! It's Saturday!,5.0
...,...
MbH BM,1.0
Lottery Ticket Checker - Florida Results & Lotto,1.0
DS Creator 2.0,1.0


Let's do the same for this

In [111]:
pd.pivot_table(playstore_reviews,index='App',values='Rating',aggfunc='mean').reset_index()

,App,Rating
0,"""i DT"" Fútbol. Todos Somos Técnicos.",4.191838
1,+Download 4 Instagram Twitter,4.500000
2,- Free Comics - Comic Apps,3.500000
3,.R,4.500000
4,/u/app,4.700000
...,...,...
9643,"뽕티비 - 개인방송, 인터넷방송, BJ방송",4.191838
9644,💎 I'm rich,3.800000
9645,"💘 WhatsLov: Smileys of love, stickers and GIF",4.600000
9646,📏 Smart Ruler ↔️ cm/inch measuring for homework!,4.000000


In [112]:
ratings = pd.pivot_table(playstore_reviews,index='App',values='Rating',aggfunc='mean').reset_index()

Types : **left, right, inner, outer, left_anti, right_anti, cross, asof**

In [122]:
sentiment_counts_ratings = pd.merge(sentiment_counts, ratings, on='App',how='inner')

In [123]:
sentiment_counts_ratings.sort_values(by='Rating',ascending=False)

,App,Negative,Neutral,Positive,Rating
398,DMV Permit Practice Test 2018 Edition,2,2,30,4.9
271,CDL Practice Test 2018 Edition,1,2,21,4.9
447,Down Dog: Great Yoga Anywhere,0,0,40,4.9
403,"DU Recorder – Screen Recorder, Video Editor, Live",11,5,42,4.8
10,30 Day Fitness Challenge - Workout at Home,2,2,27,4.8
...,...,...,...,...,...
89,Allegiant,0,1,2,3.1
416,Delta Dental,20,11,9,3.0
56,Acorn TV: World-class TV from Britain and Beyond,7,2,14,3.0
115,Anthem Anywhere,56,36,38,2.7


In [124]:
sentiment_counts_ratings.sort_values(by='Positive',ascending=False)

,App,Negative,Neutral,Positive,Rating
468,Duolingo: Learn Languages Free,238,42,1400,4.7
769,Helix Jump,348,36,1254,4.2
238,Bowmasters,714,144,1014,4.7
269,"CBS Sports App - Scores, News, Stats & Watch Live",336,216,984,4.3
302,Calorie Counter - MyFitnessPal,260,170,845,4.6
...,...,...,...,...,...
454,Draw In,1,0,0,4.2
430,Discover Mobile,1,0,0,4.6
727,HD Camera,0,1,0,4.3
453,Draw A Stickman,1,0,0,3.4


1. Do higher-rated apps have more positive reviews? yes


In [126]:
pd.crosstab(playstore_reviews['Category'],playstore_reviews['Sentiment']).sort_values(by='Positive',ascending=False)

Sentiment,Negative,Neutral,Positive
Category,,,
GAME,7270,814,11041
FAMILY,1559,463,3888
HEALTH_AND_FITNESS,577,487,3444
SPORTS,821,482,2201
PRODUCTIVITY,596,381,2144
DATING,697,530,1971
TRAVEL_AND_LOCAL,760,517,1824
PHOTOGRAPHY,485,287,1709
MEDICAL,362,315,1534


2. Which category has the most positive sentiment? GAME